### PR TITLE
add merkle tree fuzzer

### PR DESCRIPTION
**Pull request by daphneherlambda**
This commit add a fuzzer comparing to implementations for merkle tree, to run this fuzzer you need to have installed honggfuzz cause is the tool used to run it, then use de command `cargo hfuzz run merkle_tree`

### CODE REVIEW

File: merkle_patricia_tree/merkle_tree/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+        shrink,
+        expand,
+        add_sub,
+        set,
+        swap,
+        copy,
+        inter_splice,
+        insert_rand,
+        overwrite_rand,
+        // Mutator::byte_repeat_overwrite,
```
Comment (igaray) [inline merkle_patricia_tree/merkle_tree/src/main.rs:65]: Are these comments left over or are they trying to explain what the previous values mean?

File: merkle_patricia_tree/merkle_tree/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    let mut vec = vec![];
+
+    vec.push(input.to_vec());
+
+    let mut rng = StdRng::seed_from_u64(1234567);
+
+    for _ in 0..6 {
+        // Pick a random mutation strategy
+        let sel = rng.gen_range(0..STRATEGIES.len() - 1);
+        //let sel = rand_number_in_range(0, STRATEGIES.len() - 1).clone();
```
Comment (igaray) [inline merkle_patricia_tree/merkle_tree/src/main.rs]: This commented line looks like it was left over, if it's stale it would be better to remove it

**Review by igaray — CHANGES_REQUESTED** [review]
Could you add the instructions you left in the PR comments on how to setup and run inside a README.md file in the test directory?

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 1 above)
Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs:65]: They are to indicate future strategies to be implemented, for now they are merely to be a guide to the future.

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 2 above)
Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs]: it was a left over, its deleted now

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 1 above)
Comment (igaray) [reply merkle_patricia_tree/merkle_tree/src/main.rs:65]: Then just a comment indicating that is fine

File: merkle_patricia_tree/merkle_tree/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+        // things that insert into the input know that they should
+        // just insert at 0.
+        0
+    }
+}
+
+// generates a random number within a range
+fn rand_number_in_range(min: usize, max: usize) -> usize {
+
+    let mut rng = StdRng::seed_from_u64(1234567);
```
Comment (MegaRedHand) [inline merkle_patricia_tree/merkle_tree/src/main.rs]: This isn't random, as we are initializing it again on each function call ([see this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=12e06291ce4e0ba85a3b87072217fbb7)). An easy way to make it truly random (although still reproducible) would be to have the `rng` be a parameter that's initialized only once.

File: merkle_patricia_tree/merkle_tree/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+  
+    // Pick a random offset to remove data at
+    let offset = rand_offset(input);
+    // Compute the number of bytes we could remove from this offset
+    let can_remove = input.len() - offset;
+
+    // Compute a maximum number of bytes to remove
+    let max_remove = if rand_number_in_range(0,15) != 0 {
+        // 15 in 16 chance of removing at most 16 bytes, this limits the
+        // amount we remove in the most common case
```
Comment (MegaRedHand) [inline merkle_patricia_tree/merkle_tree/src/main.rs]: `rand_number_in_range(0,15)` has a 1 in 15 chance of returning 0, as it doesn't include 15 in the range (i.e. it goes from 0 to 14, for a total of 15 numbers to choose from).

File: merkle_patricia_tree/merkle_tree/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+    } else {
+        new_len = core::cmp::min(len, MAX_INPUT_SIZE - input.len());
+    }
+
+    // Create an interator to splice into the input
+    let rep = core::iter::repeat(b'\0').take(new_len);
+
+    // Expand at `dest` with `rep`, making room for the copy
+    let mut input_vec = input.to_vec();
+    input_vec.splice(dest..dest, rep);
```
Comment (MegaRedHand) [inline merkle_patricia_tree/merkle_tree/src/main.rs]: `dest..dest` is an empty range. I think you meant to use an inclusive range here (`dest..=dest`).

**Review by MegaRedHand — COMMENTED** [review]
Left some comments about rngs and ranges.

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 5 above)
Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs]: fixed the comment

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 4 above)
Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs]: fixed the comment

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 3 above)
Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs]: changed the rng generation

File: merkle_patricia_tree/merkle_tree/src/main.rs (hunk 5 above)
Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs]: done

File: merkle_patricia_tree/merkle_tree/src/main.rs [hunk 6]
```diff
@@ -209,9 +204,9 @@ fn add_sub(input: &mut [u8]) -> Vec<u8> {
     // signed integer
     let intsize = match remain {
         1..=1                => 1,
-        2..=3                => 1 << rand_number_in_range(0,1),
-        4..=7                => 1 << rand_number_in_range(0,2),
-        8..=core::usize::MAX => 1 << rand_number_in_range(0,3),
+        2..=3                => 1 << rand_number_in_range(0,1, seed),
```
Comment (MegaRedHand) [inline merkle_patricia_tree/merkle_tree/src/main.rs]: This can only return `0`, as `0..1` (i.e. `[0, 1)`) only contains the number `0`.

Comment (daphneherlambda) [reply merkle_patricia_tree/merkle_tree/src/main.rs]: fixed!

**Review by MegaRedHand — APPROVED** [review]

**Review by igaray — APPROVED** [review]

**Pull request merged**